hydromt_imod/imod.py

```diff
@@ -64,71 +66,316 @@ def __init__(
+        like: xr.DataArray,
+        method: str = "bilinear",
+    ):
+        if not ((value is None) ^ (fn is None)):
+            raise ValueError(
+                "Either provide a constant or a filename."
+                f"constant: {value}, filename: {fn}"
+            )
+
+        elif value is not None:
+            return self._prepare_constant(value, like)
+        elif fn is not None:
+            return self._prepare_raster(fn, like, method=method)
+
+    def _prepare_constant(self, value, like):
+        return xr.full_like(like, fill_value=value)
+
+    def _prepare_raster(self, fn, like, method="bilinear"):
+        da = self.data_catalog.get_rasterdataset(
+            fn,
+            geom=self.region,
+            buffer=2,
+            variables=fn,
+            single_var_as_array=True,
+        )
+        return da.raster.reproject_like(like, method=method)
+
     # setup methods
-    def setup_grid(self, region, nlay, nrow, ncol):
+    def setup_grid(self, region: dict, res: float, crs: int = "utm"):
         super().setup_region(region=region)
+        geom = self.geoms["region"]
+
+        if "bbox" in region.keys():
+            bbox = region["bbox"]
+        else:
+            bbox = geom.total_bounds
+
+        crs = hydromt.gis_utils.parse_crs(crs, bbox=bbox)
+
+        # Force geom and region to user-defined crs for model
+        # to circumvent hardcoded crs=4326 for bbox
+
+        if crs != geom.crs:
+            geom = geom.to_crs(crs)
+            self.geoms["region"] = geom
+
         # define grid
         # calculate shape & coords
         # idomain = xr.DataArray(np.ones(shape, dtype=int), coords=coords, dims=dims)
         # TODO mask with region?
         # set grid (coordinates only or dummy data)
-        # self.set_grid(idomain, 'idomain')
-
-    def setup_dis(self, top_fn, bottom_fn):
-        # NOTE: perhaps this and the above setup_mesh method can be combined
-        # read data
-        # top = self.data_catalog.get_rasterdataset(top_fn, geom=self.region, buffer=5,)
-        # bottom = self.data_catalog.get_rasterdataset(bottom_fn, geom=self.region, buffer=5,)
-        # resample to grid ??
-        # ds_dis = imod.mf6.StructuredDiscretization(
-        #     top=top, bottom=bottom, idomain=self.grid['idomain']
-        # )
-        # add data to grid
-        # self.set_grid(ds_dis, 'dis')
-        pass
+        dx, dy = res, res
+        xmin, ymin, xmax, ymax = geom.total_bounds
+        coords = imod.util._xycoords((xmin, xmax, ymin, ymax), (dx, -dy))
+        coords = {"y": coords["y"], "x": coords["x"]}
+        grid_2d = hydromt.raster.full(coords, crs=geom.crs)
 
-    def setup_recharge(self):
-        # # read data
-        # da_recharge = self.data_catalog.get_rasterdataset(recharge_fn, geom=self.region, buffer=5,)
+        self.set_grid(grid_2d, "grid_2d")
 
-        # # transform data
-        # da_recharge = imod.regrid(da_recharge)
+    def setup_dem(self, hydrography_fn):
```

Comment:
You could maybe consider merging setup_dem, setup_dis in one function? Might be clear for your user than you are kind of setting a top, bottom and number of layers. Would avoid dependency of setup_dis on setup_dem as well.

In a way what we classically do in hydromt is is one setup_* function to read one (or two) dataset(s) and derive from it all the parameters we need (eg from dem, also get slope, flow direction etc for wflow). So a setup method can easily add more than one grid